File: src/filecleaner/plan.py
"""Cleanup plans: review-then-apply, with nothing changing in between.

``fclean clean`` without ``--apply`` shows a dry run; historically a later
``--apply`` re-scanned and could quarantine something that appeared in the
meantime and was therefore never reviewed. A *plan file* closes that gap:

    fclean clean --save-plan plan.json      # review this exact list
    fclean apply plan.json                  # move exactly these, nothing else

On apply every item is re-validated: it must still exist, still be the same
kind (file/dir), and — for files — have the same size and mtime it had when
the plan was written. Anything that changed is skipped and reported.
"""

from __future__ import annotations

import contextlib
import json
import os
import stat as stat_mod
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from filecleaner import __version__
from filecleaner import scanner as scanner_mod
from filecleaner.models import Candidate, ScanResult, Skipped

PLAN_FORMAT_VERSION = 1


class PlanError(Exception):
    """The plan file is unreadable or was written by an incompatible version."""


@dataclass
class CleanupPlan:
    created_at: str
    scan_roots: list[Path]
    candidates: list[Candidate]
    tool_version: str = __version__
    format_version: int = PLAN_FORMAT_VERSION
    note: str = ""
    extra: dict[str, Any] = field(default_factory=dict)

    @property
    def total_size(self) -> int:
        return sum(c.size_bytes for c in self.candidates)

    def to_dict(self) -> dict[str, Any]:
        return {
            "format_version": self.format_version,
            "tool_version": self.tool_version,
            "created_at": self.created_at,
            "note": self.note,
            "scan_roots": [str(r) for r in self.scan_roots],
            "total_size_bytes": self.total_size,
            "candidate_count": len(self.candidates),
            "candidates": [c.to_dict() for c in self.candidates],
        }

    @classmethod
    def from_scan(cls, result: ScanResult, *, note: str = "") -> CleanupPlan:
        return cls(
            created_at=datetime.now(UTC).isoformat(timespec="seconds"),
            scan_roots=list(result.scan_roots),
            candidates=list(result.candidates),
            note=note,
        )

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> CleanupPlan:
        version = data.get("format_version")
        if version != PLAN_FORMAT_VERSION:
            raise PlanError(f"unsupported plan format_version {version!r} (expected {PLAN_FORMAT_VERSION})")
        try:
            candidates = [Candidate.from_dict(c) for c in data["candidates"]]
            roots = [Path(str(r)) for r in data.get("scan_roots", [])]
        except (KeyError, TypeError, ValueError) as exc:
            raise PlanError(f"malformed plan file: {exc}") from exc
        return cls(
            created_at=str(data.get("created_at", "")),
            scan_roots=roots,
            candidates=candidates,
            tool_version=str(data.get("tool_version", "")),
            format_version=int(version),
            note=str(data.get("note", "")),
        )


def save_plan(plan: CleanupPlan, path: Path) -> None:
    path = path.expanduser()
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_name(path.name + ".tmp")
    tmp.write_text(json.dumps(plan.to_dict(), indent=2) + "\n", encoding="utf-8")
    with contextlib.suppress(OSError):
        os.chmod(tmp, 0o600)
    os.replace(tmp, path)


def load_plan(path: Path) -> CleanupPlan:
    path = path.expanduser()
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except OSError as exc:
        raise PlanError(f"cannot read {path}: {exc}") from exc
    except json.JSONDecodeError as exc:
        raise PlanError(f"{path} is not valid JSON: {exc}") from exc
    if not isinstance(data, dict):
        raise PlanError(f"{path}: expected a JSON object")
    return CleanupPlan.from_dict(data)


def revalidate(plan: CleanupPlan) -> tuple[list[Candidate], list[Skipped]]:
    """Split a plan into items still safe to apply and items that changed."""
    fresh: list[Candidate] = []
    stale: list[Skipped] = []
    for cand in plan.candidates:
        try:
            st = os.lstat(cand.path)
        except FileNotFoundError:
            stale.append(Skipped(str(cand.path), "no longer exists"))
            continue
        except OSError as exc:
            stale.append(Skipped(str(cand.path), f"cannot stat: {exc.strerror or exc}"))
            continue

        if stat_mod.S_ISLNK(st.st_mode):
            stale.append(Skipped(str(cand.path), "is now a symlink"))
            continue
        is_dir = stat_mod.S_ISDIR(st.st_mode)
        if is_dir != cand.is_dir:
            stale.append(Skipped(str(cand.path), "changed between file and directory"))
            continue
        if is_dir:
            # A directory candidate's size/mtime describe everything inside
            # it, recursively — re-derive the same numbers the same way the
            # scanner did, so content added/changed/removed since the plan
            # was written is caught, not just changes to the top folder.
            # (An adversarially backdated mtime with a coincidentally
            # unchanged total size could in principle slip through; real
            # cache/log directories never do this, and it's still bounded
            # by every other safety check purge/quarantine already apply.)
            size, newest_inside = scanner_mod.dir_stats(cand.path)
            changed = size != cand.size_bytes or max(st.st_mtime, newest_inside) > cand.mtime + 1e-6
        else:
            changed = st.st_size != cand.size_bytes or abs(st.st_mtime - cand.mtime) > 1e-6
        if changed:
            stale.append(Skipped(str(cand.path), "modified since the plan was written"))
            continue
        fresh.append(cand)
    return fresh, stale

File: native/fclean-walk/src/plan.rs
//! `plan`: the file `fclean clean --save-plan` writes and `fclean apply`
//! reads back. Saved, loaded and revalidated here; *applying* one moves
//! files, and is not ported.
//!
//! The file is the rollback path between the two implementations, so either
//! must accept what the other writes. This one is the stricter reader: it
//! takes what Python writes, and refuses hand-edited values that Python
//! would coerce (`"12"` for a size, `1` for a flag, `true` for the version).

use std::fs;
use std::io::ErrorKind;
use std::os::unix::fs::PermissionsExt;
use std::path::Path;
use std::time::SystemTime;

use serde::Deserialize;
use serde_json::{Map, Value};

use crate::pyjson::{dumps, dumps_ascii, J};
use crate::pypath::{expanduser, normalise};
use crate::report::{candidate_json, locate, scan, strerror, NativeRequest, Scan, Setup};
use crate::rules::repr;
use crate::{dir_stats, fail, mtime_nanos, pytime, st_mtime, start_pool};

const FORMAT_VERSION: i64 = 1;
/// How far a file's time may be from the plan's and still be the same file.
const MTIME_TOLERANCE: f64 = 1e-6;

pub struct Planned {
    pub path: String,
    pub size: i64,
    pub is_dir: bool,
    pub mtime: f64,
    pub rule_id: String,
    pub category: String,
    pub risk: String,
}

impl Planned {
    /// `Candidate.to_dict`.
    fn to_json(&self) -> J {
        J::dict([
            ("path", J::str(&self.path)),
            ("size_bytes", J::Int(self.size)),
            ("is_dir", J::Bool(self.is_dir)),
            ("mtime", J::Float(self.mtime)),
            ("rule_id", J::str(&self.rule_id)),
            ("category", J::str(&self.category)),
            ("risk", J::str(&self.risk)),
        ])
    }
}

fn malformed(what: impl std::fmt::Display) -> String {
    format!("malformed plan file: {what}")
}

/// `Candidate.from_dict`. A missing key is reported in Python's words — a
/// `KeyError` prints as the key's `repr`.
fn planned(item: &Value) -> Result<Planned, String> {
    let item = item.as_object().ok_or_else(|| malformed("a candidate is not an object"))?;
    let field = |key: &str| item.get(key).ok_or_else(|| malformed(repr(key)));
    let text = |key: &str| field(key)?.as_str().map(str::to_owned).ok_or_else(|| malformed(format!("{key} is not text")));
    Ok(Planned {
        path: normalise(&text("path")?),
        size: field("size_bytes")?.as_i64().ok_or_else(|| malformed("size_bytes is not a whole number"))?,
        is_dir: field("is_dir")?.as_bool().ok_or_else(|| malformed("is_dir is not true or false"))?,
        mtime: field("mtime")?.as_f64().ok_or_else(|| malformed("mtime is not a number"))?,
        rule_id: text("rule_id")?,
        category: text("category")?,
        risk: if item.contains_key("risk") { text("risk")? } else { "low".to_owned() },
    })
}

/// `CleanupPlan.from_dict`, as far as `revalidate` needs it.
fn from_dict(data: &Map<String, Value>) -> Result<Vec<Planned>, String> {
    let version = data.get("format_version");
    if version.and_then(Value::as_f64) != Some(FORMAT_VERSION as f64) {
        let shown = match version {
            None | Some(Value::Null) => "None".to_owned(),
            Some(Value::String(text)) => repr(text),
            Some(other) => other.to_string(),
        };
        return Err(format!("unsupported plan format_version {shown} (expected {FORMAT_VERSION})"));
    }
    if data.get("scan_roots").is_some_and(|roots| !roots.is_array()) {
        return Err(malformed("scan_roots is not a list"));
    }
    let candidates = data.get("candidates").ok_or_else(|| malformed(repr("candidates")))?;
    candidates.as_array().ok_or_else(|| malformed("candidates is not a list"))?.iter().map(planned).collect()
}

/// `load_plan`.
pub fn load_plan(path: &str) -> Result<Vec<Planned>, String> {
    let bytes = fs::read(path).map_err(|err| format!("cannot read {path}: {err}"))?;
    let text = String::from_utf8(bytes).map_err(|_| format!("cannot read {path}: it is not UTF-8"))?;
    let data: Value = serde_json::from_str(&text).map_err(|err| format!("{path} is not valid JSON: {err}"))?;
    from_dict(data.as_object().ok_or_else(|| format!("{path}: expected a JSON object"))?)
}

/// `revalidate`: what is still exactly what was reviewed, and what is not,
/// with the reason. A folder's size and time describe everything inside it,
/// so they are taken again the way the scan took them.
pub fn revalidate(plan: Vec<Planned>) -> (Vec<Planned>, Vec<(String, &'static str, Option<String>)>) {
    let mut fresh = Vec::new();
    let mut stale = Vec::new();
    for candidate in plan {
        let meta = match fs::symlink_metadata(&candidate.path) {
            Ok(meta) => meta,
            Err(err) if err.kind() == ErrorKind::NotFound => {
                stale.push((candidate.path, "no longer exists", None));
                continue;
            }
            Err(err) => {
                let why = err.raw_os_error().map_or_else(|| err.to_string(), strerror);
                stale.push((candidate.path, "cannot stat: ", Some(why)));
                continue;
            }
        };
        let own = mtime_nanos(&meta);
        let reason = if meta.file_type().is_symlink() {
            Some("is now a symlink")
        } else if meta.is_dir() != candidate.is_dir {
            Some("changed between file and directory")
        } else if meta.is_dir() {
            let (size, newest) = dir_stats(&candidate.path, own);
            let changed = i64::try_from(size) != Ok(candidate.size) || st_mtime(newest.max(own)) > candidate.mtime + MTIME_TOLERANCE;
            changed.then_some("modified since the plan was written")
        } else {
            let changed = i64::try_from(meta.len()) != Ok(candidate.size) || (st_mtime(own) - candidate.mtime).abs() > MTIME_TOLERANCE;
            changed.then_some("modified since the plan was written")
        };
        match reason {
            Some(reason) => stale.push((candidate.path, reason, None)),
            None => fresh.push(candidate),
        }
    }
    (fresh, stale)
}

#[derive(Deserialize)]
struct CheckRequest {
    #[serde(flatten)]
    setup: Setup,
    plan: String,
}

/// `{"fresh": [...], "stale": [...]}`: what `fclean apply PLAN` would move,
/// and what it would leave and why — without moving anything.
pub fn check_main(input: &str) {
    let request: CheckRequest =
        serde_json::from_str(input).unwrap_or_else(|err| fail(format!("bad plan-check-json request: {err}")));
    let (env, _) = locate(&request.setup);
    let path = expanduser(&request.plan, &env.home).unwrap_or_else(|err| fail(err));
    let plan = load_plan(&path).unwrap_or_else(|err| fail(err));
    start_pool(request.setup.threads);

    let (fresh, stale) = revalidate(plan);
    let skipped = |(path, reason, detail): &(String, &str, Option<String>)| {
        J::dict([("path", J::str(path)), ("reason", J::str(format!("{reason}{}", detail.as_deref().unwrap_or(""))))])
    };
    let report = J::dict([
        ("fresh", J::List(fresh.iter().map(Planned::to_json).collect())),
        ("stale", J::List(stale.iter().map(skipped).collect())),
    ]);
    println!("{}", dumps(&report));
}

#[derive(Deserialize)]
struct SaveRequest {
    #[serde(flatten)]
    scan: NativeRequest,
    /// `fclean clean --save-plan PATH`.
    save_plan: String,
    #[serde(default)]
    note: String,
    /// The clock, when two plans have to be compared.
    created_at: Option<String>,
}

/// `CleanupPlan.from_scan(...).to_dict()`.
fn plan_json(found: &Scan, note: &str, created_at: &str) -> J {
    J::dict([
        ("format_version", J::Int(FORMAT_VERSION)),
        ("tool_version", J::str(env!("CARGO_PKG_VERSION"))),
        ("created_at", J::str(created_at)),
        ("note", J::str(note)),
        ("scan_roots", J::strings(&found.roots)),
        ("total_size_bytes", J::UInt(found.scanned.candidates.iter().map(|hit| hit.size).sum())),
        ("candidate_count", J::UInt(found.scanned.candidates.len() as u64)),
        ("candidates", J::List(found.scanned.candidates.iter().map(|hit| candidate_json(hit, &found.owners[hit.walk])).collect())),
    ])
}

/// `save_plan`: written beside itself and moved into place, readable by its
/// owner alone — it lists what is on the disk.
fn save_plan(plan: &J, path: &str) -> std::io::Result<()> {
    if let Some(parent) = Path::new(path).parent().filter(|parent| !parent.as_os_str().is_empty()) {
        fs::create_dir_all(parent)?;
    }
    let tmp = format!("{path}.tmp");
    fs::write(&tmp, dumps_ascii(plan) + "\n")?;
    let _ = fs::set_permissions(&tmp, fs::Permissions::from_mode(0o600));
    fs::rename(&tmp, path)
}

/// The scan, saved as a plan: `fclean clean --save-plan PATH` without the
/// dry-run table. Writes the one file it is asked to write.
pub fn save_main(input: &str) {
    let request: SaveRequest =
        serde_json::from_str(input).unwrap_or_else(|err| fail(format!("bad plan-save request: {err}")));
    let (env, _) = locate(&request.scan.setup);
    let path = expanduser(&request.save_plan, &env.home).unwrap_or_else(|err| fail(err));
    let created_at = request.created_at.clone().unwrap_or_else(|| {
        // datetime.now(UTC).isoformat(timespec="seconds"): cut, not rounded.
        let seconds = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).map_or(0, |since| since.as_secs() as i64);
        format!("{}+00:00", pytime::isoformat(seconds, 0))
    });

    let found = scan(&request.scan);
    save_plan(&plan_json(&found, &request.note, &created_at), &path).unwrap_or_else(|err| fail(format!("cannot write {path}: {err}")));
    let report = J::dict([("plan", J::str(&path)), ("candidate_count", J::UInt(found.scanned.candidates.len() as u64))]);
    println!("{}", dumps(&report));
}

#[cfg(test)]
mod tests {
    use super::*;

    fn plan(candidate: &str) -> Map<String, Value> {
        let text = format!(r#"{{"format_version": 1, "scan_roots": ["/r"], "candidates": [{candidate}]}}"#);
        serde_json::from_str::<Value>(&text).unwrap().as_object().unwrap().clone()
    }

    const WHOLE: &str = r#"{"path": "/r//a/", "size_bytes": 3, "is_dir": false, "mtime": 5, "rule_id": "x", "category": "C"}"#;

    #[test]
    fn what_python_writes_is_read_back() {
        let read = from_dict(&plan(WHOLE)).unwrap();
        assert_eq!((read[0].path.as_str(), read[0].size, read[0].mtime, read[0].risk.as_str()), ("/r/a", 3, 5.0, "low"));
    }

    #[test]
    fn a_missing_key_is_named_the_way_python_names_it() {
        let without_size = WHOLE.replace(r#""size_bytes": 3, "#, "");
        assert_eq!(from_dict(&plan(&without_size)).err().unwrap(), "malformed plan file: 'size_bytes'");
    }

    #[test]
    fn values_python_would_coerce_are_refused() {
        for (from, to) in [(r#""size_bytes": 3"#, r#""size_bytes": "3""#), (r#""is_dir": false"#, r#""is_dir": 0"#)] {
            assert!(from_dict(&plan(&WHOLE.replace(from, to))).is_err(), "{to}");
        }
        let mut other_version = plan(WHOLE);
        other_version.insert("format_version".to_owned(), Value::from(2));
        assert_eq!(from_dict(&other_version).err().unwrap(), "unsupported plan format_version 2 (expected 1)");
    }
}
